CSS Custom Highlights optimization for collecting markedText

jesxilin opened the pull request:
#### 85b79f4d702c0d5ca90a0d576c9cce404a6b3b5e
<pre>
CSS Custom Highlights optimization for collecting markedText
<a href="https://bugs.webkit.org/show_bug.cgi?id=259797">https://bugs.webkit.org/show_bug.cgi?id=259797</a>
rdar://113356533

Reviewed by Megan Gardner.

Cleaned up implementation for finding insertion index for custom highlights,
taking into account set priorities and insertion order.

* Source/WebCore/rendering/MarkedText.cpp:
(WebCore::MarkedText::collectForHighlights):

Canonical link: <a href="https://commits.webkit.org/266601@main">https://commits.webkit.org/266601@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1c301b62ca1ef075652183b4583a6c9e768d7dfd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/14155 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/14468 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/14805 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/15892 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/13416 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/14256 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16979 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/14549 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/16088 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/14329 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14896 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12005 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/16611 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12179 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/12766 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/19784 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13258 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12930 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16133 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/13473 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/11334 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/12752 "Built successfully") | | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3447 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17087 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13316 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/rendering/MarkedText.cpp:
```diff
@@ -145,18 +145,14 @@ Vector<MarkedText> MarkedText::collectForHighlights(const RenderText& renderer,
-                                markedTexts.insert(0, { highlightStart, highlightEnd, MarkedText::Type::Highlight, nullptr, highlightName, currentPriority });
+                            // Gets the first place such that it > currentPriority.
+                            auto it = std::upper_bound(markedTexts.begin(), markedTexts.end(), currentPriority, [](const auto& m1, const auto& m2) {
+                                return m1 > m2.priority;
```
What do you think about making clear that `m1` is `currentPriority` and `m2` is a `MarkedText`?

Plus the first arg is a POD type so we don't need a const ref.

```suggestion
                            auto it = std::upper_bound(markedTexts.begin(), markedTexts.end(), currentPriority, [](const auto currentPriority, const auto& markedText) {
                                return currentPriority > markedText.priority;
```

aprotyas reviewed (approved):
Looks good overall. Suggesting some changes for readability

rr-codes commented on Source/WebCore/rendering/MarkedText.cpp:
```diff
@@ -145,18 +145,14 @@ Vector<MarkedText> MarkedText::collectForHighlights(const RenderText& renderer,
-                            if (!wasInserted)
-                                markedTexts.insert(0, { highlightStart, highlightEnd, MarkedText::Type::Highlight, nullptr, highlightName, currentPriority });
+                            // Gets the first place such that it > currentPriority.
+                            auto it = std::upper_bound(markedTexts.begin(), markedTexts.end(), currentPriority, [](const auto& m1, const auto& m2) {
```
```suggestion
                            auto it = std::upper_bound(markedTexts.begin(), markedTexts.end(), currentPriority, [](const auto& targetMarkedTextPriority, const auto& markedText) {
```

megangardner reviewed (approved):
With the changes that Abrar and Richard suggested.

jesxilin committed:
CSS Custom Highlights optimization for collecting markedText
https://bugs.webkit.org/show_bug.cgi?id=259797
rdar://113356533

Reviewed by Megan Gardner.

Cleaned up implementation for finding insertion index for custom highlights,
taking into account set priorities and insertion order.

* Source/WebCore/rendering/MarkedText.cpp:
(WebCore::MarkedText::collectForHighlights):

Canonical link: https://commits.webkit.org/266601@main
